cpp/csp/python/PyEngine.h

```diff
@@ -54,6 +55,24 @@ class CSPIMPL_EXPORT PyEngine final: public PyObject
     Engine    * m_engine;
 };
 
+// Generic engine shutdown function for push-type adapters
+template<typename T>
+PyObject * PyEngine_shutdown( T * self, PyObject * args, PyObject * kwargs )
+{
+    CSP_BEGIN_METHOD;
+
+    PyObject * pyException;
+    if( !PyArg_ParseTuple( args, "O", &pyException ) )
+        return NULL;
+
+    if( !PyExceptionInstance_Check( pyException ) )
+       CSP_THROW( TypeError, "Expected Exception object as argument for shutdown_engine: got " << Py_TYPE( pyException ) -> tp_name );
```

Comment:
one small concern here, if someone mis-uses the call and passes say a string instead of an exception, this raise would end up not shutting down the engine either...
maybe we should still call engine shutdown in this case ( and then throw ) but this needs to be tested